federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProviderFactory.java

```diff
@@ -340,29 +340,10 @@ public void onCreate(KeycloakSession session, RealmModel realm, ComponentModel m
         // CN is typically used as RDN for Active Directory deployments
         if (ldapConfig.getRdnLdapAttribute().equalsIgnoreCase(LDAPConstants.CN)) {
 
-            if (usernameLdapAttribute.equalsIgnoreCase(LDAPConstants.CN)) {
-
-                // For AD deployments with "cn" as username, we will map "givenName" to first name
-                mapperModel = KeycloakModelUtils.createComponentModel("first name", model.getId(), UserAttributeLDAPStorageMapperFactory.PROVIDER_ID,LDAPStorageMapper.class.getName(),
-                        UserAttributeLDAPStorageMapper.USER_MODEL_ATTRIBUTE, UserModel.FIRST_NAME,
-                        UserAttributeLDAPStorageMapper.LDAP_ATTRIBUTE, LDAPConstants.GIVENNAME,
-                        UserAttributeLDAPStorageMapper.READ_ONLY, readOnly,
-                        UserAttributeLDAPStorageMapper.ALWAYS_READ_VALUE_FROM_LDAP, alwaysReadValueFromLDAP,
-                        UserAttributeLDAPStorageMapper.IS_MANDATORY_IN_LDAP, "true");
-                realm.addComponentModel(mapperModel);
-
-            } else {
+            if (!usernameLdapAttribute.equalsIgnoreCase(LDAPConstants.CN)) {
                 if (editMode == UserStorageProvider.EditMode.WRITABLE) {
 
-                    // For AD deployments with "sAMAccountName" as username and writable, we need to map "cn" as username as well (this is needed so we can register new users from KC into LDAP) and we will map "givenName" to first name.
-                    mapperModel = KeycloakModelUtils.createComponentModel("first name", model.getId(), UserAttributeLDAPStorageMapperFactory.PROVIDER_ID,LDAPStorageMapper.class.getName(),
-                            UserAttributeLDAPStorageMapper.USER_MODEL_ATTRIBUTE, UserModel.FIRST_NAME,
-                            UserAttributeLDAPStorageMapper.LDAP_ATTRIBUTE, LDAPConstants.GIVENNAME,
-                            UserAttributeLDAPStorageMapper.READ_ONLY, readOnly,
-                            UserAttributeLDAPStorageMapper.ALWAYS_READ_VALUE_FROM_LDAP, alwaysReadValueFromLDAP,
-                            UserAttributeLDAPStorageMapper.IS_MANDATORY_IN_LDAP, "true");
-                    realm.addComponentModel(mapperModel);
-
+                    // For AD deployments with "sAMAccountName" as username and writable, we need to map "cn" as username as well (this is needed so we can register new users from KC into LDAP).
```

Comment:
Just wondering if this mapper is about making sure there is a value set for CN when creating users given that `firstName` and `lastName` might be optional.

If so, I think the mapper for CN might involve a specific logic to check whether the first and last names are optional and if so, fallback to the username value. Instead of mapping to a single attribute. This logic should not be specific to AD, I think, and other vendors will have the same mapper. 

This is kinda what the `FullNameLDAPStorageMapper` does but it does make much sense to put the fullname into the username on KC side.